model/jpa/src/main/java/org/keycloak/authorization/jpa/entities/PolicyEntity.java

```diff
@@ -59,7 +56,7 @@
                 @NamedQuery(name="findPolicyIdByResource", query="select p from PolicyEntity p inner join fetch p.resources r left join fetch p.scopes s inner join fetch p.associatedPolicies a where p.resourceServer.id = :serverId and (r.resourceServer.id = :serverId and r.id = :resourceId)"),
                 @NamedQuery(name="findPolicyIdByScope", query="select pe from PolicyEntity pe left join fetch pe.resources r inner join fetch pe.scopes s inner join fetch pe.associatedPolicies a where pe.resourceServer.id = :serverId and exists (select p.id from ScopeEntity s inner join s.policies p where s.resourceServer.id = :serverId and (p.resourceServer.id = :serverId and p.type = 'scope' and s.id in (:scopeIds) and p.id = pe.id))"),
                 @NamedQuery(name="findPolicyIdByResourceScope", query="select pe from PolicyEntity pe inner join fetch pe.resources r inner join fetch pe.scopes s inner join fetch pe.associatedPolicies a where pe.resourceServer.id = :serverId and exists (select p.id from ScopeEntity s inner join s.policies p where s.resourceServer.id = :serverId and (p.resourceServer.id = :serverId and p.type = 'scope' and s.id in (:scopeIds) and p.id = pe.id)) and exists (select p.id from ResourceEntity r inner join r.policies p where r.resourceServer.id = :serverId and (p.resourceServer.id = :serverId and p.id = pe.id and p.type = 'scope' and r.id in (:resourceId)))"),
-                @NamedQuery(name="findPolicyIdByNullResourceScope", query="select pe from PolicyEntity pe left join fetch pe.resources r inner join fetch pe.scopes s inner join fetch pe.associatedPolicies a where pe.resourceServer.id = :serverId and exists (select p.id from ScopeEntity s inner join s.policies p where s.resourceServer.id = :serverId and (p.resourceServer.id = :serverId and p.id = pe.id and p.type = 'scope' and s.id in (:scopeIds))) and pe.resources is empty"),
+                @NamedQuery(name="findPolicyIdByNullResourceScope", query="select pe from PolicyEntity pe left join fetch pe.resources r left join fetch pe.configKeys ck inner join fetch pe.scopes s inner join fetch pe.associatedPolicies a where pe.resourceServer.id = :serverId and exists (select p.id from ScopeEntity s inner join s.policies p where s.resourceServer.id = :serverId and (p.resourceServer.id = :serverId and p.id = pe.id and p.type = 'scope' and s.id in (:scopeIds))) and pe.resources is empty and 'defaultResourceType' not member of pe.configKeys"),
```

Comment:
We realized that, by adding a `resourceType` to `ScopePermissionRepresentation`, we are implying that such Permissions refer to a set of resources. Scope permissions which do not reference an specific `resource` are involved in all Resource Permission Evaluation requesting the specified scope.  The query `findPolicyIdByNullResourceScope` looks for Scope Permissions which do not refer to specific resources. When a `ScopePermission` has a `defaultResourceType` defined, it should not be retrieved by this query. E.g.:

* Resources:
  - t1 (Scopes=read,write, Type=t)
  - t2 (Scopes=read,write, Type=t)
  - x1 (Scopes=read,write, Type=x)
  - x2 (Scopes=read,write, Type=x)
* Permissions:
  - ScopePermission -> t:read (Scope=read, ResourceType=t)
  - ScopePermission -> any:read (Scope=read, ResourceType=null)

Without the proposed changes, the permission `t:read` would be retrieved by the `DefaultPolicyEvaluator` when evaluating permissions for the resource `x1` since it doesn't reference an specific resource (It references a `resourceType` instead). Only the permission `any:read` should be involved in the `x1` permission evaluation since `t:read`only applies to resources of type `t`. 


We've included a new field `configKeys` since JPA does not support queries like `'defaultResourceType' not member of pe.config` being `pe.config` a `Map` collection. This could also be achieved with an in-memory filtering but we think the JPA approach is a better solution. Do you think is there a better approach?

Reply:
This scenario is covered in the test we provided

Reply:
I'm a bit concerned about this change. I'm afraid it could impact performance?

It is indeed better than filtering the results once they are fetched from the database. Really not sure ...

Reply:
We found a better approach to this query which doesn't need the addition of the `configKeys`  field and only uses the existing `config` map. We could leave the query as follows:

select pe from PolicyEntity pe 
&nbsp;&nbsp;&nbsp;left join fetch pe.resources r 
&nbsp;&nbsp;&nbsp;**left join fetch pe.config c**
&nbsp;&nbsp;&nbsp;inner join fetch pe.scopes s 
&nbsp;&nbsp;&nbsp;inner join fetch pe.associatedPolicies a 
where pe.resourceServer.id = :serverId 
&nbsp;&nbsp;&nbsp;and exists (select p.id from ScopeEntity s inner join s.policies p where s.resourceServer.id = :serverId and (p.resourceServer.id = :serverId and p.id = pe.id and p.type = 'scope' and s.id in (:scopeIds)))
&nbsp;&nbsp;&nbsp;and pe.resources is empty 
&nbsp;&nbsp;&nbsp;**and not exists (select pec from pe.config pec where KEY(pec) = 'defaultResourceType')**

what do you think?

Reply:
Hi @pedroigor, we've seen that you resolved this conversation. Do you have any thoughts about the change proposal we made related to this query? Maybe we could update the PR with the proposed changes (the ones which do not require the `configKeys` field)? 
I reopened the conversation, let me know if you want to continue this conversation in the main thread of the PR :)

Reply:
@alvarogimenez Sorry, back to this one now. I think we can start without the configKeys field.

Reply:
Done :)